spec/edpm-openstack-service-architecture.md

```diff
@@ -0,0 +1,326 @@
+start the OpenStack service binary. While parts of this is visible to the human
+deployer. E.g. kubernetes Pods, Deployments, ConfigMaps, VolumeMounts. The
+human deployer only needs to understand the `CustomServiceConfig` text fields
+to be able to pass through the desired configuration to an OpenStack service.
+
+As OpenStack services are running also on the EDPM Nodes there is also a need
+to pass through human defined configuration to these services as well.
+
+On the EDPM side Ansible roles are used to provision and configure services.
+Ansible has its own configuration passing mechanism, Ansible variables. It is
+technically possible to pass through oslo.config options as Ansible variables
+to configure the OpenStack services.
+
+However it means that the same oslo.config configuration interface of the
+OpenStack service are represented in two different formats. On the podified
+control plane side it is an oslo.config snippet in the `CustomServiceConfig`
+field. While on the EDPM side Ansible variables. The former does not involve
+any translation. The naming and structure of the config options are exactly the
+same in the `CustomServiceConfig` field as in the actual oslo.config file of
+the OpenStack binary. In the other hand the latter applies a name mangling to
+avoid name clashes between different oslo.config sections and between different
+services using the same Ansible inventory to pass through variables. Moreover
+the latter requires explicit work in the Ansible roles every time a new config
+variable is added.
+
+For example, enabling debug logging for the OVN metadata agent on the EDPM node
+requires the following oslo.config file to be provided to the service:
+```ini
+  [DEFAULT]
+  debug = True
+```
+
+With the use of the `CustomServiceConfig` the following needs to be added to
+the CR representing the given OVN metadata agent:
+```yaml
+customServiceConfig: |
+  [DEFAULT]
+  debug = True
+```
+
+With ansible variables the following needs to be added to the CR that triggers
+deployment of the given OVN metadata agent:
+```yaml
+ansibleVars: |
+  edpm_ovn_metadata_agent_metadata_agent_DEFAULT_debug: true
+```
+
+## Proposed change
+
+Keep the existing DataPlane, DataPlaneRole, and DataPlaneNode CRDs to represent
+and coordinate phase -1 and phase 0 of the EDPM deployment. But decouple
+phase 1 from the existing DataPlane model. This means that the DataPlane CRDs
+will not represent and therefore won't directly deploy any OpenStack services
+on the EDPM node. These CRDs still expose the status of the phase -1 and
+phase 0 deployment, so other operators can decide when to start deploying
+OpenStack service to the EDPM node.
+
+Define CRDs representing each OpenStack service type that needs to be run on
+the EDPM node (e.g. NovaExternalCompute, OVNMetadataAgent, etc). These CRDs
+are defined in and therefore reconciled by the respective service operator
+(e.g. nova-operator, ovn-operator). These OpenStack service CRDs will be
```

Comment:
nit: the OVNMetadataAgent woudl actully be reconsiled by the neutron-operator since its not actually part of ovn its part of neutron, just like the ovn ml2 driver is part of neutron. Its the neutron agent that implement the metadata proxy for the ovn backend.